File: modules/webrtc-core/src/main/java/com/shterneregen/securelan/webrtc/runtime/RtcFileLogger.java
package com.shterneregen.securelan.webrtc.runtime;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.Instant;

public final class RtcFileLogger {
    private static final Object LOCK = new Object();
    private static final String LOG_FILE_PROPERTY = "securelan.rtc.logFile";
    private static final Path DEFAULT_LOG_FILE = Path.of("logs", "rtc.log");

    private RtcFileLogger() {
    }

    public static void warn(String message) {
        write("WARN", message, null);
    }

    public static void error(String message) {
        write("ERROR", message, null);
    }

    public static void error(String message, Throwable error) {
        write("ERROR", message, error);
    }

    private static void write(String level, String message, Throwable error) {
        synchronized (LOCK) {
            try {
                Path logFile = logFile();
                Path parent = logFile.getParent();
                if (parent != null) {
                    Files.createDirectories(parent);
                }

                Files.writeString(
                        logFile,
                        format(level, message, error),
                        StandardCharsets.UTF_8,
                        StandardOpenOption.CREATE,
                        StandardOpenOption.APPEND
                );
            } catch (Throwable ignored) {
                // Logging must never break realtime media flow.
            }
        }
    }

    private static Path logFile() {
        String configuredPath = System.getProperty(LOG_FILE_PROPERTY);
        if (configuredPath == null || configuredPath.isBlank()) {
            return DEFAULT_LOG_FILE;
        }
        return Path.of(configuredPath.trim());
    }

    private static String format(String level, String message, Throwable error) {
        StringBuilder builder = new StringBuilder()
                .append(Instant.now())
                .append(' ')
                .append(level)
                .append(" [rtc] ")
                .append(message == null ? "" : message)
                .append(System.lineSeparator());

        if (error != null) {
            StringWriter buffer = new StringWriter();
            error.printStackTrace(new PrintWriter(buffer));
            builder.append(buffer);
        }

        return builder.toString();
    }
}

File: modules/webrtc-core/src/main/kotlin/com/shterneregen/securelan/webrtc/runtime/RtcFileLogger.kt
package com.shterneregen.securelan.webrtc.runtime

import java.io.PrintWriter
import java.io.StringWriter
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.time.Instant

object RtcFileLogger {
    private val lock = Any()
    private const val LOG_FILE_PROPERTY = "securelan.rtc.logFile"
    private val DEFAULT_LOG_FILE: Path = Path.of("logs", "rtc.log")

    @JvmStatic
    fun warn(message: String?) {
        write("WARN", message, null)
    }

    @JvmStatic
    fun error(message: String?) {
        write("ERROR", message, null)
    }

    @JvmStatic
    fun error(message: String?, error: Throwable?) {
        write("ERROR", message, error)
    }

    private fun write(level: String, message: String?, error: Throwable?) {
        synchronized(lock) {
            try {
                val logFile = logFile()
                val parent = logFile.parent
                if (parent != null) {
                    Files.createDirectories(parent)
                }

                Files.writeString(
                    logFile,
                    format(level, message, error),
                    StandardCharsets.UTF_8,
                    StandardOpenOption.CREATE,
                    StandardOpenOption.APPEND,
                )
            } catch (_: Throwable) {
                // Logging must never break realtime media flow.
            }
        }
    }

    private fun logFile(): Path {
        val configuredPath = System.getProperty(LOG_FILE_PROPERTY)
        if (configuredPath.isNullOrBlank()) {
            return DEFAULT_LOG_FILE
        }
        return Path.of(configuredPath.trim())
    }

    private fun format(level: String, message: String?, error: Throwable?): String {
        val builder = StringBuilder()
            .append(Instant.now())
            .append(' ')
            .append(level)
            .append(" [rtc] ")
            .append(message ?: "")
            .append(System.lineSeparator())

        if (error != null) {
            val buffer = StringWriter()
            error.printStackTrace(PrintWriter(buffer))
            builder.append(buffer)
        }

        return builder.toString()
    }
}
